src/codegen/compile.ml

```diff
@@ -5795,23 +5796,94 @@ module FuncDec = struct
       compile_unboxed_const (E.add_fun_ptr env (E.built_in env reject_name)) ^^
       get_cb_index
 
+  let closures_to_self_reply_reject_callbacks env ts =
+    let reply_name = "@self_callback<" ^ Typ_hash.typ_hash (Type.Tup ts) ^ ">" in
+    Func.define_built_in env reply_name ["env", I32Type] [] (fun env ->
+        let (set_arr, get_arr) = new_local env "arr" in
+        message_start env (Type.Shared Type.Write) ^^
+        (* Look up closure *)
+        let (set_closure, get_closure) = new_local env "closure" in
+        G.i (LocalGet (nr 0l)) ^^
+        ClosureTable.recall env ^^
+        set_arr ^^ get_arr ^^
+        Arr.load_field 0l ^^ (* get the reply closure *)
+        set_closure ^^
+        get_closure ^^
+
+        (* Deserialize reply arguments  *)
+        Serialization.deserialize env ts ^^
+
+        get_closure ^^
+        Closure.call_closure env (List.length ts) 0 ^^
+
+        message_cleanup env (Type.Shared Type.Write)
+      );
+
+    let reject_name = "@self_reject_callback" in
+    Func.define_built_in env reject_name ["env", I32Type] [] (fun env ->
+        message_start env (Type.Shared Type.Write) ^^
+        (* Look up closure *)
+        let (set_closure, get_closure) = new_local env "closure" in
+        G.i (LocalGet (nr 0l)) ^^
+        ClosureTable.recall env ^^
+        Arr.load_field 1l ^^ (* get the reject closure *)
+        set_closure ^^
+        get_closure ^^
+        (* Synthesize value of type `Text`, the error message
+           (The error code is fetched via a prim)
+        *)
+        IC.error_value env ^^
+
+        get_closure ^^
+        Closure.call_closure env 1 0 ^^
+
+        message_cleanup env (Type.Shared Type.Write)
+      );
+
+    (* The (above) upper half of this function must not depend on the
+       get_future, get_k and get_r parameters, so hide them from above (cute trick) *)
+    fun get_future get_k get_r ->
+      let (set_cb_index, get_cb_index) = new_local env "cb_index" in
+      (* store the tuple away, future_array_index = 2, keep in sync with rts/closure_table.rs *)
+      Arr.lit env [get_k; get_r; get_future] ^^
```

Comment:
```suggestion
      Arr.lit env [get_k; get_r; get_future] ^^
```
Continuing from my comment above: is this the only difference from the cloned code? If so, maybe refactor the original as:
```
fun get_array ->
      let (set_cb_index, get_cb_index) = new_local env "cb_index" in
      (* store the tuple away, future_array_index = 2, keep in sync with rts/closure_table.rs *)
      get_array ^^
...
```
and have the caller pass in the appropriate get_array code (constructed at the call site)

If possible, this will avoid duplicating some gnarly code (and share more code to boot).

Reply:
Yep, the `ic_call_threaded` now deals with code holes that are customisable. This way I could unify remote calls and self calls, which only differ in argument passing and how the continuations are pushed.